rpc/erigon.go

```diff
@@ -768,3 +770,339 @@ type Eth1InternalTransactionWithPosition struct {
+	} else {
+		b.ExcessBlobGas = nil
+	}
+
+	if string(tempData.Uncles) == "null" || string(tempData.Uncles) == "" {
+		b.Uncles = []*geth_types.Block{}
+	} else if tempData.Uncles[0] == '[' {
+		// unmarshal as an array of type geth_types.Block
+		if err := json.Unmarshal(tempData.Uncles, &b.Uncles); err != nil {
+			// if it failed then unmarshal as an array of strings
+			var uncleHashes []string
+			if err := json.Unmarshal(tempData.Uncles, &uncleHashes); err != nil {
+				return fmt.Errorf("expected an array for uncles, but got: %s", tempData.Uncles)
+			}
+			// convert string hashes to empty Block refs
+			for i := range uncleHashes {
+				b.Uncles[i] = &geth_types.Block{}
+			}
+		}
+	} else {
+		return fmt.Errorf("expected an array for uncles, but got: %s", tempData.Uncles)
+	}
+
+	return nil
+}
+
+func (client *ErigonClient) GetBlocksByBatch(blocksChan chan *types.Eth1Block, erigonClient *ErigonClient) ([]*types.Eth1Block, *types.GetBlockTimings, error) {
+	startTime := time.Now()
+	defer func() {
+		metrics.TaskDuration.WithLabelValues("rpc_el_get_blocks_by_batch").Observe(time.Since(startTime).Seconds())
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	timings := &types.GetBlockTimings{}
+
+	var blockDetails []*types.Eth1Block
+	var batchCall []geth_rpc.BatchElem
+	batchCallNums := 3
+
+	if blocksChan == nil {
+		fmt.Printf("block channel is empty")
+	}
+
+	for block := range blocksChan {
+		batchCall = append(batchCall, geth_rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{block.Number, true},
+			Result: new(json.RawMessage),
+		})
+
+		batchCall = append(batchCall, geth_rpc.BatchElem{
+			Method: "eth_getBlockReceipts",
+			Args:   []interface{}{block.Number},
+			Result: new([]geth_types.Receipt),
+		})
+
+		batchCall = append(batchCall, geth_rpc.BatchElem{
+			Method: "trace_block",
+			Args:   []interface{}{block.Number},
+			Result: new([]ParityTraceResult),
+		})
+	}
+
+	if len(batchCall) == 0 {
+		return blockDetails, timings, nil
+	}
+
+	err := erigonClient.rpcClient.BatchCallContext(ctx, batchCall)
+	if err != nil {
+		logger.Errorf("error while batch calling rpc, error: %s", err)
+	}
+
+	for i := 0; i < len(batchCall)/batchCallNums; i++ {
+		blockResult := batchCall[i*batchCallNums].Result.(*json.RawMessage)
+		receiptsResult := batchCall[i*batchCallNums+1].Result.(*[]geth_types.Receipt)
+		tracesResults := batchCall[i*batchCallNums+2].Result.(*[]ParityTraceResult)
+
+		var blockResponse BlockResponse
+		err := json.Unmarshal([]byte(*blockResult), &blockResponse)
+		if err != nil {
+			fmt.Printf("\n errror while unmarshalling block results %s\n", err)
+		}
+
+		blockDetail, timings := client.processBlockResult(blockResponse, timings)
+		blockDetail, timings = client.processReceiptsAndTraces(blockDetail, *receiptsResult, *tracesResults, timings)
+		blockDetails = append(blockDetails, blockDetail)
+	}
+
+	return blockDetails, timings, nil
```

Comment:
there is something wrong with `timings`. It's always overridden by the last block you proceed.
Moreover we need to think about it in the context of that function. I think it use to record the time taken to retrieve a block (through 3 requests). In your case it's now one request for N blocks.
I would say just remove it, and if it's deemed important create a new timing type for batch request.